services/autoscaling/src/simcore_service_autoscaling/utils_docker.py

```diff
@@ -2,163 +2,219 @@
+        return (
+            f"{match.group(1)} --availability=drain {match.group(2)} {match.group(3)}"
+        )
+    raise RuntimeError(
+        f"expected docker '{_DOCKER_SWARM_JOIN_RE}' command not found: received {decoded_stdout}!"
+    )
+
+
+@retry(
+    stop=stop_after_delay(_TIMEOUT_WAITING_FOR_NODES),
+    before_sleep=before_sleep_log(logger, logging.WARNING),
+    after=after_log(logger, logging.ERROR),
+    wait=wait_fixed(5),
+)
+async def wait_for_node(node_name: str) -> Node:
     async with aiodocker.Docker() as docker:
+        list_of_nodes = await docker.nodes.list(filters={"name": node_name})
+    if not list_of_nodes:
+        raise TryAgain
+    return parse_obj_as(Node, list_of_nodes[0])
 
-        services = await docker.services.list()
-        for service in services:
-            tasks = service.tasks()
-            for task in tasks:
-                if (
-                    task["Status"]["State"] == "running"
-                    and task["NodeID"] in nodes_ids
-                    and (
-                        reservations := task["Spec"]
-                        .get("Resources", {})
-                        .get("Reservations")
-                    )
-                ):
-                    tasks_resources.append(
-                        {"ID": task["ID"], **_get_reserved_resources(reservations)}
-                    )
-
-                elif (
-                    task["Status"]["State"] == "pending"
-                    and task["Status"]["Message"] == "pending task scheduling"
-                    and "insufficient resources on" in task["Status"]["Err"]
-                    and (
-                        reservations := task["Spec"]
-                        .get("Resources", {})
-                        .get("Reservations")
-                    )
-                ):
-                    tasks_pending_resources.append(
-                        {"ID": task["ID"], **_get_reserved_resources(reservations)}
-                    )
-
-        total_tasks_cpus = 0
-        total_tasks_ram = 0
-        for task in tasks_resources:
-            total_tasks_cpus += task["CPU"]
-            total_tasks_ram += task["RAM"]
-
-        for task in tasks_pending_resources:
-            total_pending_tasks_cpus += task["CPU"]
-            total_pending_tasks_ram += task["RAM"]
-
-        return TasksResources(
-            total_cpus_running_tasks=total_tasks_cpus,
-            total_ram_running_tasks=total_tasks_ram,
-            total_cpus_pending_tasks=total_pending_tasks_cpus,
-            total_ram_pending_tasks=total_pending_tasks_ram,
-            count_tasks_pending=len(tasks_pending_resources),
+
+async def tag_node(node: Node, *, tags: dict[str, str], available: bool) -> None:
+    async with aiodocker.Docker() as docker:
+        assert node.ID  # nosec
+        assert node.Version  # nosec
+        assert node.Version.Index  # nosec
+        assert node.Spec  # nosec
+        assert node.Spec.Role  # nosec
```

Comment:
Same question here, are these also always set?
I don't feel happy about these asserts.

Reply:
this is directly coming from the pydantic generated model. nothing to worry about.